File: algorithms/M/Move Pieces to Obtain a String/Move Pieces to Obtain a String.cpp
class Solution {
public:
    

    bool canChange(string start, string target) {
        
        
        vector<pair<char,int >> v;
        for(int i =0 ;i  < start.size();i++)
        {
            if(start[i] == 'L' || start[i] == 'R')
            {
                v.push_back({start[i],i});
            }
        }
        
        int i = 0;
        for(int j = 0;j < target.size();j++)
        {
            if(target[j] == 'L' || target[j] == 'R') // lets find its corresponding Element in start
            {
                if(i >= v.size() || v[i].first != target[j]) // Not found or Different found
                {
                    return false;
                }else 
                {
                   
                    if(v[i].first == 'L' && v[i].second < j) // Condition a unsatisfied
                        return false;
                     if(v[i].first == 'R' && v[i].second > j) // Condition b unsatisfied
                        return false;
                     else 
                        i++; // all good we move ahead in start;
                }
               
                
            }
        }
        
        return i == v.size();// all elements accounted for in start
    }
};

File: scripts/algorithms/M/Move Pieces to Obtain a String/Move Pieces to Obtain a String.js
// Runtime: 151 ms (Top 47.54%) | Memory: 47.4 MB (Top 60.66%)
/**
 * @param {string} start
 * @param {string} target
 * @return {boolean}
 */
var canChange = function(start, target) {
    let i = 0
    let j = 0
    const len = start.length

    while(i <= len && j <= len) {
        while(i < len && start[i] === '_') i++
        while(j < len && target[j] === '_') j++

        if(i === len && j === len) return true
        if(i === len || j === len) return false

        if(start[i] !== target[j]) return false
        if(start[i] === 'L' && j > i) return false
        if(start[i] === 'R' && j < i) return false
        i++
        j++
    }

    return false
};
